### DISCUSSION TITLE
Configuring a relationship between 2 tables that reference others

### DISCUSSION THREAD

**Discussion by jvanasco** [sqlalchemy/sqlalchemy#8600] [Usage Questions]
In the following example, I have two core tables, `Foo` and `Bar`, alongside a relationship table `Foo2Bar`.

The project grew, and a new association table was added `Foo2BarOrBiz`.  This table  tracks some data about relationships between `Foo` with either `Bar` or `Biz`.  

The following code raises this error when the mappers are configured, as I try to define the relationship between `Foo2Bar` and `Foo2BarOrBiz`:

```
Traceback (most recent call last):
  File "/Users/jvanasco/Desktop/sqla_example_remote.py", line 82, in <module>
    configure_mappers()
  File "/Users/jvanasco/dev-env/cliquedin-3.10.2/lib/python3.10/site-packages/sqlalchemy/orm/mapper.py", line 3418, in configure_mappers
    _configure_registries(_all_registries(), cascade=True)
  File "/Users/jvanasco/dev-env/cliquedin-3.10.2/lib/python3.10/site-packages/sqlalchemy/orm/mapper.py", line 3448, in _configure_registries
    _do_configure_registries(registries, cascade)
  File "/Users/jvanasco/dev-env/cliquedin-3.10.2/lib/python3.10/site-packages/sqlalchemy/orm/mapper.py", line 3487, in _do_configure_registries
    mapper._post_configure_properties()
  File "/Users/jvanasco/dev-env/cliquedin-3.10.2/lib/python3.10/site-packages/sqlalchemy/orm/mapper.py", line 1941, in _post_configure_properties
    prop.init()
  File "/Users/jvanasco/dev-env/cliquedin-3.10.2/lib/python3.10/site-packages/sqlalchemy/orm/interfaces.py", line 231, in init
    self.do_init()
  File "/Users/jvanasco/dev-env/cliquedin-3.10.2/lib/python3.10/site-packages/sqlalchemy/orm/relationships.py", line 2140, in do_init
    self._generate_backref()
  File "/Users/jvanasco/dev-env/cliquedin-3.10.2/lib/python3.10/site-packages/sqlalchemy/orm/relationships.py", line 2476, in _generate_backref
    self._add_reverse_property(self.back_populates)
  File "/Users/jvanasco/dev-env/cliquedin-3.10.2/lib/python3.10/site-packages/sqlalchemy/orm/relationships.py", line 2079, in _add_reverse_property
    raise sa_exc.ArgumentError(
sqlalchemy.exc.ArgumentError: Foo2Bar.foo_2_bar_or_biz and back-reference Foo2BarOrBiz.foo_2_bar are both of the same direction symbol('ONETOMANY').  Did you mean to set remote_side on the many-to-one side ?
```

I am not sure why `remote_side` is being suggested here - as this is not a self-referential usage.  Dropping the `back_populates` avoids the exception, but worries me.  I've tried several strategies to configure the relationship, but none seem to work correctly - so my understanding of this part of SQLAlchemy is clearly deficient.  Any help would be greatly appreciated.

```
from sqlalchemy import CheckConstraint
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import Unicode
from sqlalchemy.orm import configure_mappers
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import relationship


Base = declarative_base()


class Foo(Base):
    __tablename__ = "foo"
    id = Column(Integer, primary_key=True)
    name = Column(Unicode(255), nullable=False)


class Bar(Base):
    __tablename__ = "bar"
    id = Column(Integer, primary_key=True)
    name = Column(Unicode(255), nullable=False)


class Foo2Bar(Base):
    __tablename__ = "foo_2_bar"
    foo_id = Column(Integer, ForeignKey("foo.id"), primary_key=True)
    bar_id = Column(Integer, ForeignKey("bar.id"), primary_key=True)

    foo_2_bar_or_biz = relationship(
        "Foo2BarOrBiz",
        primaryjoin=(
            "and_("
            "Foo2Bar.foo_id==foreign(Foo2BarOrBiz.foo_id),"
            "Foo2Bar.bar_id==foreign(Foo2BarOrBiz.bar_id),"
            ")"
        ),
        uselist=False,
        back_populates="foo_2_bar",
    )


class Foo2BarOrBiz(Base):
    __tablename__ = "foo_2_bar_or_biz"
    __table_args__ = (
            CheckConstraint(
            """
            (
                (bar_id IS NULL and biz IS NOT NULL)
                OR
                (bar_id IS NOT NULL and biz IS NULL)
            )
            """,
            name="check_bar_or_biz",
        ),
)
    id = Column(Integer, primary_key=True)
    foo_id = Column(Integer, ForeignKey("foo.id"), nullable=False)
    bar_id = Column(Integer, ForeignKey("bar.id"), nullable=True)
    biz = Column(Unicode(255), nullable=True)
    foo_2_bar = relationship(
        "Foo2Bar",
        primaryjoin=(
            "and_("
            "Foo2BarOrBiz.foo_id==foreign(Foo2Bar.foo_id),"
            "Foo2BarOrBiz.bar_id==foreign(Foo2Bar.bar_id),"
            ")"
        ),
        uselist=False,
        back_populates="foo_2_bar_or_biz",
    )


engine = create_engine(
    "sqlite://",
    echo=True,
)
Base.metadata.create_all(engine)

configure_mappers()
```

**Comment by zzzeek** [comment]
you have the foreign() conditions switched on the two relationships which is why they both think they are onetomany.   two relationships that are bidirectional would always have the identical columns as foreign(), these are the cols that would have FOREIGN KEY on them.   from a relationship writing perspective these are the columns that receive the value.  so if you were to assign to Foo2BarOrBiz.foo_2_bar, the ORM would have to copy from Foo2BarOrBiz.foo_id to Foo2Bar.foo_id.   This is contradictory towards if you were to assign to Foo2Bar.foo_2_bar_or_biz, it would by copying Foo2Bar.foo_id to Foo2BarOrBiz.foo_id.

these two relationships also seem like they have to be viewonly, since foo_id/bar_id on both are FOREIGN KEY to foo and bar, so you would never write to these columns by assigning to "foo_2_bar" or "foo_2_bar_or_biz", they already match foo.id / bar.id.  if they are both viewonly then back_populates is mostly meaningless.

**Reply by jvanasco** [reply]
Yes, they ARE both viewonly on the real application.  I just do the back_populates to ensure I have a corresponding relationship set up on each side.  

I guess I can either use the same join condition, or comment out the `back_populates`.  Both break our style guide a bit, but work.

Thanks a ton, Mike!

**Reply by zzzeek** [reply]
these are both what I guess we could call "cheater" relationships since these cols are kind of overlapping with some other table, so they get viewonly=True and from there you usually dont want to use back_populates, since nothing is being "populated".
